Add From React to Effect blog post (#809)
Effect-TS/website: content/blog/0045-from-react-to-effect.mdx

  })

const printMessages = (messages: number) => Effect.gen(function*() {
  for (let i = 0; i < messages; i++) {
    yield* print(`message: ${i}`)
  }
})

Effect.runPromise(printMessages(10))
```

<br />
<a href="https://effect.website/play#3c5c45b473a4" target="_blank">Open in Playground</a>

Similarly to how React components compose together we've now seen that also Effects can easily be composed, in the same way as we were able to raise an error from a component we can just as easily raise an error from an Effect and recover it from a parent Effect:

```ts
import { Effect } from "effect"

const print = (message: string) =>
  Effect.sync(() => {
    console.log(message)
  })

class InvalidRandom extends Error {
  message = "Invalid Random Number"
}

const printOrFail = Effect.gen(function*() {
  if (Math.random() > 0.5) {
    yield* print("Hello World")
  } else {
    yield* Effect.fail(new InvalidRandom())
  }
})

const program = printOrFail.pipe(
  Effect.catchAll((e) => print(`Error: ${e.message}`)),
  Effect.repeatN(10)
)

Effect.runPromise(program)
```

<br />
<a href="https://effect.website/play#a35378f5d2ab" target="_blank">Open in Playground</a>

> The above code randomly fails with an error that we recover from a parent Effect, in this case the recovery logic is to print the error message, differently from `React` we can see here that errors are type safe, in fact in our `catchAll` we know that `e` is of type `InvalidRandom`, this is possible because Effect uses inference to understand which error cases your program may encounter and represent those cases in its type, you can check the type of `printOrFail` and you'll see: `Effect<void, InvalidRandom, never>` which means that this Effect will return `void` if successful but may fail for `InvalidRandom`. You can compose Effects that can fail for different reasons and your final Effect will list all possible errors in a union, so you'd get for example `Effect<number, InvalidRandom | NetworkError | ..., never>`.

Reviewer: ```suggestion
The above code will randomly fail with a `InvalidRandom` error, which we then recover from a parent Effect using `Effect.catchAll`. In this case, the recovery logic is to simply log the error message to the console.

However, what separates Effect from `React` is that errors are 100% type safe - within our `Effect.catchAll`, we know that `e` is of type `InvalidRandom`. This is possible because Effect uses type inference to understand which error cases your program may encounter and represent those cases in its type. If you check the type of `printOrFail`, you will see: 

`Effect<void, InvalidRandom, never>` 

which means that this Effect will return `void` if successful but may also fail with an `InvalidRandom` error. 

When you compose Effects that may fail for different reasons, your final Effect will list all possible errors in a union, so you would see something like the following in the type: 

`Effect<number, InvalidRandom | NetworkError | ..., never>`.
```

> An Effect can represent **any** piece of code, let it be a console.log statement, a fetch call, a database query or a computation, in fact Effect is capable of executing both async and sync code in a unified model, escaping the issue of <a href="https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/" target="_blank">function coloring</a> (having different types for async or sync).

Reviewer: ```suggestion
An Effect can represent **any** piece of code, let it be a `console.log` statement, a fetch call, a database query or a computation. Effect is also fully capable of executing both synchronous and asynchronous code in a unified model, escaping the issue of <a href="https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/" target="_blank">function coloring</a> (i.e. having different types for async or sync).
```

> Just like React components are able to access context provided by a parent component also Effects can access context provided from a parent Effect, let's see how:

Reviewer: ```suggestion
Just like React components can access context provided by a parent component, Effects can also access context provided from a parent Effect. Let's see how:
```

```ts
import { Context, Effect } from "effect"

const print = (message: string) =>
  Effect.sync(() => {
    console.log(message)
  })

class ContextualData extends Context.Tag("ContextualData")<ContextualData, number>() {}

const printFromContext = Effect.gen(function*() {
  const n = yield* ContextualData

  yield* print(`Contextual Data is: ${n}`)
})

const program = printFromContext.pipe(
  Effect.provideService(ContextualData, 100)
)

Effect.runPromise(program)
```

<br />
<a href="https://effect.website/play#88051688ce37" target="_blank">Open in Playground</a>

> Differently from React we don't have to provide a default implementation for our context, Effect will keep track of the requirements in its third type parameter and it will prohibit execution of an Effect that doesn't have all requirements fullfilled, you can check the type of `printFromContext` and will see `Effect<void, never, ContextualData>` which means that this Effect will return `void` upon success and that it needs `ContextualData` in order to become executable.

Reviewer: ```suggestion
What separates Effect from React here is that we are not forced to provide a default implementation for our context. Effect keeps track of all requirements of our program in its third type parameter, and will prohibit execution of an Effect that does not have all requirements fulfilled.

If you check the type of `printFromContext`, you will see:

`Effect<void, never, ContextualData>`

which means that this Effect will return `void` upon success, does not fail with any expected errors, and requires `ContextualData` in order to become executable.
```

## Conclusion

> We can see that Effect and React use basically the same model, both are about making composable descriptions that are later executed by a runtime, only the domain is different while React focuses on UI Effect focuses on general purpose code.

Reviewer: ```suggestion
We can see that Effect and React share essentially the same underlying model - both libraries are about making composable descriptions of a program that can later be executed by a runtime. Only the domain is different - React focuses on building user interfaces while Effect focuses on creating general purpose programs.
```

This is only an introduction and Effect provides much more than what's shown here, this includes features such as:

- Concurrency
- Retry Scheduling
- Telemetry
- Metrics
- Logging

And much more.

> If you're curious your next read should be [the docs](https://effect.website/docs/introduction) and your next watch should be the [Effect Beginner Workshop](https://www.youtube.com/watch?v=Lz2J1NBnHK4).

Reviewer: ```suggestion
If you're curious about Effect, please checkout our [docs](https://effect.website/docs/introduction) as well as the [Effect Beginner Workshop](https://www.youtube.com/watch?v=Lz2J1NBnHK4).
```
